### PR TITLE
[#49493] Add a `WorkPackageRole` type

### PR DESCRIPTION
**Pull request by aaron-contreras**
## Milestones
1. A new Role type will become available (`WorkPackageRole`) in order to capture/assign roles to users and groups appropriately for Work-Package-specific permissions in order to grant access to OpenProject on a Work-Package-specific basis.

2. Ability to specify/determine which role types permissions are assignable to. This will help simplify the form implementation for Role Creation.

3. Seed the three default Work Package roles and corresponding permissions that will come pre-packaged with OpenProject
  - Editor
  - Commenter
  - Viewer 

## Notes
Work Package: https://community.openproject.org/work_packages/49493

### CODE REVIEW

File: app/seeders/basic_data/base_role_seeder.rb [hunk 1]
```diff
@@ -74,7 +72,9 @@ def permissions(value)
       when Array
         value
       when :all_assignable_permissions
-        Roles::CreateContract.new(Role.new, nil).assignable_permissions.map(&:name)
+        Roles::CreateContract.new(self.class.model_class.new, nil)
+                             .assignable_permissions
+                             .map(&:name)
```
Comment (aaron-contreras) [inline app/seeders/basic_data/base_role_seeder.rb]: I don't know if this is a bug or intended but this would always assign all `assignable_permissions` for Project Roles even if a `GlobalRole` is being seeded. @cbliard

File: spec/seeders/basic_data/project_role_seeder_spec.rb [hunk 2]
```diff
@@ -103,42 +105,15 @@
...
-    end
-  end
-
   context 'with permissions: :all_assignable_permissions' do
     let(:data_hash) do
       YAML.load <<~SEEDING_DATA_YAML
         roles:
-        - reference: :role_project_admin
-          name: Project admin
-          global: true
```
Comment (aaron-contreras) [inline spec/seeders/basic_data/project_role_seeder_spec.rb:139]: @cbliard Following up on the `assignable_permissions` comment. This was intended to be a global role, but based on the spec it was expected to be assigned all `assignable_permissions` for the `Role` type.

Now that I've split these specs into `project_role_seeder_spec` and `global_role_seeder_spec` I've added an expectation on the `global_role_seeder_spec` that the assigned roles to the created role, only gets the `assignable_permissions` for `GlobalRole` types.

File: app/seeders/common.yml [hunk 3]
```diff
@@ -240,111 +240,115 @@ open_color_palette:
...
+        - :view_changesets
+        - :view_team_planner
+    - reference: :default_role_project_admin
+      t_name: Project admin
+      position: 5
+      permissions: :all_assignable_permissions
+  global:
+    - reference: :default_role_project_creator_and_staff_manager
+      t_name: Staff and projects manager
+      global: true
```
Comment (cbliard) [inline app/seeders/common.yml]: `global: true` is not useful anymore.

File: spec/seeders/basic_data/project_role_seeder_spec.rb (hunk 2 above)
Comment (cbliard) [reply spec/seeders/basic_data/project_role_seeder_spec.rb:139]: I think I did a mistake originally here. As it is the "Project admin" role, it should not have been `global: true`.
Thanks for having fixed it!

File: spec/support/seeders.rb [hunk 4]
```diff
@@ -48,7 +54,7 @@ def add_needed_seeders_for(seeder_class, needed_acc)
   shared_let(:needed_seeders_keys) do
     # warning: there is one limitation: RoleSeeder will only pick 'roles' key,
     # and not 'modules_permissions' key.
-    needed_seeders.map { _1.try(:seed_data_model_key) }.compact
```
Comment (cbliard) [inline spec/support/seeders.rb]: I do not like introducing exceptions. It can then evolve to accidental complexity. The idea was to keep a 1-1 relation between the key name in the seed files and the class name of the record. This way I do not have to think too much to lookup the right key given a model name.

Instead of having `roles.project`, `roles.global`, and `roles.work_package` , why not use `project_roles`, `global_roles`, and `work_package_roles` directly? This way the role seeders would work as any other seeder.

What do you think about it?

File: spec/seeders/basic_data/global_role_seeder_spec.rb [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+              global: true
+              position: 1
+              permissions: :all_assignable_permissions
+      SEEDING_DATA_YAML
+    end
+
+    it 'gives all assignable permissions to the role' do
+      expect(GlobalRole.find_by(name: 'Project admin').permissions)
+        .to match_array(Roles::CreateContract.new(GlobalRole.new, nil).assignable_permissions.map { _1.name.to_sym })
+    end
```
Comment (cbliard) [inline spec/seeders/basic_data/global_role_seeder_spec.rb:83]: I think that's ok. When I run it in a console, it returns `[:add_project, :create_backup, :create_user, :manage_user, :manage_placeholder_user]`

Maybe the test can be enhanced to check that the permissions include `:add_project` and `:create_user`, and do not include `view_project` and `view_work_packages`. Just to be sure that they are global permissions and not project permissions.

(As a side note, I'm always suspicious when we assert equality using business code in the comparison: in this case, what if a bug is introduced in `Roles::CreateContract.new(GlobalRole.new, nil).assignable_permissions`? I suppose we have to trust that a unit test in `Roles::CreateContract` exists for when a `GlobalRole` is given. In that case, hardening the test would be useless. And after checking, it looks like `spec/contracts/roles/shared_contract_examples.rb` is covering it, so no need to add the "test enhancements" I suggested earlier: that would only duplicate code)

File: app/seeders/basic_data/base_role_seeder.rb [hunk 6]
```diff
@@ -43,8 +41,8 @@ def model_attributes(role_data)
 
     private
 
-    def type(global_value)
-      true?(global_value) ? 'GlobalRole' : 'Role'
+    def type
+      self.class.model_class.to_s
```
Comment (cbliard) [inline app/seeders/basic_data/base_role_seeder.rb]: `self.class` is not needed. `model_class.to_s` is enough.

File: app/seeders/basic_data/base_role_seeder.rb (hunk 1 above)
Comment (cbliard) [reply app/seeders/basic_data/base_role_seeder.rb]: That's not what I am observing. I tried it and it returned the expected global permissions.

We can have a look together. Ping me when you're available.

File: app/seeders/basic_data/base_role_seeder.rb [hunk 7]
```diff
@@ -74,7 +72,9 @@ def permissions(value)
       when Array
         value
       when :all_assignable_permissions
-        Roles::CreateContract.new(Role.new, nil).assignable_permissions.map(&:name)
+        Roles::CreateContract.new(self.class.model_class.new, nil)
```
Comment (cbliard) [inline app/seeders/basic_data/base_role_seeder.rb]: `self.class` not needed here.

File: spec/seeders/basic_data/global_role_seeder_spec.rb (hunk 5 above)
Comment (aaron-contreras) [reply spec/seeders/basic_data/global_role_seeder_spec.rb:83]: Discussed in a call

**Review by klaustopher — COMMENTED** [review]
Looked through everything. Very thorough!

File: lib/open_project/access_control/permission.rb [hunk 8]
```diff
@@ -42,6 +42,7 @@ def initialize(name,
                      hash,
                      public: false,
                      require: nil,
+                     permissible_on: [],
```
Comment (klaustopher) [inline lib/open_project/access_control/permission.rb]: I'd opt to remove the default value here and make it mandatory to add this to a  permission definiton

Comment (aaron-contreras) [reply lib/open_project/access_control/permission.rb]: Ah, yes! I was thinking about adding a warning somewhere in case someone forgets about defining it but the error message raised by removing the default is much better. :)

**Review by klaustopher — APPROVED** [review]
Awesome work!
